source/Irrlicht/CQuantNormalCache.h

```diff
@@ -68,11 +69,114 @@ class CQuantNormalCache
 		}
 	};
 
+	template<E_QUANT_NORM_CACHE_TYPE CacheType>
+	struct vector_for_cache;
+
+	template<> 
+	struct vector_for_cache<E_QUANT_NORM_CACHE_TYPE::Q_8_8_8>
+	{
+		typedef uint32_t type;
+	};
+
+	template<> 
+	struct vector_for_cache<E_QUANT_NORM_CACHE_TYPE::Q_2_10_10_10>
+	{
+		typedef uint32_t type;
+	};
+
+	template<> 
+	struct vector_for_cache<E_QUANT_NORM_CACHE_TYPE::Q_16_16_16>
+	{
+		typedef uint64_t type;
+	};
+
+	template<E_QUANT_NORM_CACHE_TYPE CacheType>
+	using vector_for_cache_t = typename vector_for_cache<CacheType>::type;
+
 public:
+	template<E_QUANT_NORM_CACHE_TYPE CacheType>
+	vector_for_cache_t<CacheType> quantizeNormal(const core::vectorSIMDf& normal)
+	{
+		uint32_t quantizationBits = 0;
+		IRR_PSEUDO_IF_CONSTEXPR_BEGIN(CacheType == E_QUANT_NORM_CACHE_TYPE::Q_2_10_10_10)
+		{
+			quantizationBits = 10u;
+		}
+		IRR_PSEUDO_IF_CONSTEXPR_END
+		IRR_PSEUDO_IF_CONSTEXPR_BEGIN(CacheType == E_QUANT_NORM_CACHE_TYPE::Q_8_8_8)
+		{
+			quantizationBits = 8u;
+		}
+		IRR_PSEUDO_IF_CONSTEXPR_END
+		IRR_PSEUDO_IF_CONSTEXPR_BEGIN(CacheType == E_QUANT_NORM_CACHE_TYPE::Q_16_16_16)
+		{
+			quantizationBits = 10u; //for sure?
+		}
+		IRR_PSEUDO_IF_CONSTEXPR_END
+
+		const auto xorflag = core::vectorSIMDu32((0x1u << quantizationBits) - 1u);
+		const auto negativeMask = normal < core::vectorSIMDf(0.0f);
+		const VectorUV uvMappedNormal = mapToBarycentric(core::abs(normal));
 
-	uint32_t quantizeNormal2_10_10_10(const core::vectorSIMDf& normal);
-	uint32_t quantizeNormal8_8_8(const core::vectorSIMDf& normal);
-	uint64_t quantizeNormal16_16_16(const core::vectorSIMDf& normal);
+		IRR_PSEUDO_IF_CONSTEXPR_BEGIN(CacheType == E_QUANT_NORM_CACHE_TYPE::Q_2_10_10_10)
+		{
+			auto found = normalCacheFor2_10_10_10Quant.find(uvMappedNormal);
+			if (found != normalCacheFor2_10_10_10Quant.end() && (found->first == uvMappedNormal))
+			{
+				const uint32_t absVec = found->second;
+				const auto vec = core::vectorSIMDu32(absVec, absVec >> quantizationBits, absVec >> quantizationBits * 2) & xorflag;
```

Comment:
careful about an expression like `absVec >> quantizationBits * 2` the bitwise operators have weird precedence, its better to write `absVec >> (quantizationBits * 2)`